### ISSUE TITLE
miniengine current updated frame

### ISSUE THREAD

**Issue by jmkinzer** [Microsoft/DirectX-Graphics-Samples#105]
In various other examples, for example:
[https://github.com/Microsoft/DirectX-Graphics-Samples/blob/master/Samples/Desktop/D3D12DynamicIndexing/src/D3D12DynamicIndexing.cpp#L593](url)
A frameindex that points to the frame currently being rendered to is maintained along with a fence to avoid running too far ahead and updating a frame still being processed by the GPU. Is g_CurrentBuffer the same concept in miniengine? I'm not sure because I don't see similar fences, I'm probably missing something.

**Comment by stanard** [comment]
MiniEngine tracks fences after each command list execution.  These fences are used for resource reclamation.  They are also used for synchronizing the Graphics and Compute command queues.  Though they usually aren't used to synchronize CPU and GPU, they sometimes get used that way during resource creation time.  (E.g. load texture data, upload texture data, wait on fence, destroy original data buffer.)

g_CurrentBuffer is used to keep track of which swap chain buffer is being rendered to.  It's important for presenting the frames in the right order, but it's not used for fencing purposes.  There is a built-in mechanism to stall, however.  When you call Present(), you have a maximum frame latency of 3.  If you have submitted 3 frames already that have yet to be presented, the driver will idle the CPU until one of the frames is displayed.  You can also try using DXGI_SWAP_CHAIN_FLAG_FRAME_LATENCY_WAITABLE_OBJECT, which let's you change the maximum frame latency and also let's you decide when the CPU should be throttled.

**Comment by jmkinzer** [comment]
Thanks for the reply! A bit of background: I've implemented something like the [dynamic indexing example](https://github.com/Microsoft/DirectX-Graphics-Samples/tree/master/Samples/Desktop/D3D12DynamicIndexing). (I use dynamic descriptors for some things but prefer that approach for the case of high instance counts). So like in the example I have N frames worth of descriptors but I'm uncertain of the best way is to ensure I'm rendering the "correct" frame of descriptors (versus updating them). I think ideally the Update() and RenderScene() would just have the current frame number as context when they call me. Does this makes any sense or is there some other solution you would recommend?

**Comment by stanard** [comment]
Yes, you can call Graphics::GetFrameCount() to determine the current frame index that the CPU is processing.  You could then somehow associate the frame index with your descriptors if you needed to.  For simplicity's sake, you could have four GPU-visible descriptor heaps (or just rotate through offsets into a single buffer), and then

Update DescriptorTable[ FrameCount % 4 ]
Set DescriptorTable[ FrameCount % 4 ]
Draw objects with DescriptorTable[ FrameCount % 4 ]

DescriptorTable[ FrameCount % 4 ] will persist for three more frames
Present() will increment FrameCount

But for even more simplicity's sake, you could pre-build a very large descriptor table with all of your descriptors.  The indices used to refer to individual resources will remain constant from frame-to-frame.  If a resource goes out of scope (i.e. is released) and thereby makes its descriptor invalid, make sure you don't use that descriptor index for the time being.  But if you stream that resource back in (like texture streaming), you would reuse that descriptor index.

I hope I didn't confuse you.  I'm not totally sure what you're intending, so please excuse me if I'm missing the point.

**Comment by jmkinzer** [comment]
Thanks James - sounds like something along the lines of Graphics::GetFrameCount() % SWAP_CHAIN_BUFFER_COUNT is what I'm after.
